driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -105,34 +121,42 @@ public ServerTuple selectServer(final ServerSelector serverSelector, final Opera
         try {
             CountDownLatch currentPhase = phase.get();
             ClusterDescription curDescription = description;
+            logServerSelectionStarted(operationContext, serverSelector, curDescription);
             ServerSelector compositeServerSelector = getCompositeServerSelector(serverSelector);
             ServerTuple serverTuple = selectServer(compositeServerSelector, curDescription);
 
-            boolean selectionFailureLogged = false;
+            boolean selectionWaitingLogged = false;
 
             long startTimeNanos = System.nanoTime();
             long curTimeNanos = startTimeNanos;
-            long maxWaitTimeNanos = getMaxWaitTimeNanos();
+            Long maxWaitTimeNanos = getMaxWaitTimeNanos();
 
             while (true) {
-                throwIfIncompatible(curDescription);
+                if (!curDescription.isCompatibleWithDriver()) {
+                    throw createAndLogIncompatibleException(operationContext, serverSelector, curDescription);
+                }
 
                 if (serverTuple != null) {
+                    logServerSelectionSucceeded(operationContext, serverTuple.getServerDescription().getAddress(), serverSelector, curDescription);
                     return serverTuple;
                 }
 
-                if (curTimeNanos - startTimeNanos > maxWaitTimeNanos) {
-                    throw createTimeoutException(serverSelector, curDescription);
+                Long remainingTimeNanos = maxWaitTimeNanos == null ? null : maxWaitTimeNanos - (curTimeNanos - startTimeNanos);
+
+                if (remainingTimeNanos != null && remainingTimeNanos <= 0) {
```

Comment:
Note how now `remainingTimeNanos <= 0` is used, while previously `curTimeNanos - startTimeNanos > maxWaitTimeNanos` was used, which was equivalent to `remainingTimeNanos < 0`. The reason we must use `<=` instead of `<` is that when `maxWaitTimeNanos` is 0, we must not wait (as per the `getServerSelectionTimeout` docs), and `<= 0` makes sure we don't wait, while `<` is not guaranteed to give us that behavior.

Reply:
The Javadoc says:

```
A value of 0 means that it will timeout immediately if no server is available.  
A negative value means to wait indefinitely.
```

The spec only says:

```
A serverSelectionTimeoutMS of zero MAY have special meaning in some drivers; 
zero's meaning is not defined in this spec, but all drivers SHOULD document the meaning of zero.
```

So this looks good to me.